### DISCUSSION TITLE
Custom actions

### DISCUSSION THREAD

**Discussion by johtso** [ahochsteger/gmail-processor#330] [Q&A] [answered]
How do you go about defining a custom action? I don't see any mention of this in the docs.

I wanted to define a simple action that would send the attachment via http to a specific URL.

**Comment by ahochsteger** [comment]
@johtso currently the only way to add custom actions is by modifying the source code and build it yourself.
May you explain your use case a bit more? Esp. how the HTTP call (I guess a POST) would look like (which headers, which content type)?
I'd be happy to accept contributions as well since this might be useful for others as well.
In case you'd want to give it a try you may have a look at the implementation of the action [`message.storeFromURL`](https://github.com/ahochsteger/gmail-processor/blob/main/src/lib/actions/MessageActions.ts#L180) which stores a document from an URL to Google Drive where the URL is typically extracted from the body of the message.

**Comment by ahochsteger** [comment, marked as answer]
@johtso I've just released version 2.11.0 which adds support for custom actions.
Have a look at [this example](https://ahochsteger.github.io/gmail-processor/docs/examples/advanced/customActions) to see how it can be used.
It would be great to get some feedback, if you are able to migrate to Gmail Processor with this new feature.

**Reply by johtso** [reply]
Oh great, I'll try and take a look! I ended up just copy and pasting the code and modifying it directly because my needs were so simple, but seems like a nice feature to add.

**Comment by johtso** [comment]
Haven't had a chance to really dig into it yet, but does this support async actions?

**Comment by ahochsteger** [comment]
@johtso, unfortunately not, since the Google Apps Script APIs are synchronous.
But maybe these resources may be helpful:
* https://stackoverflow.com/questions/53824779/asynchronous-execution-of-a-function-app-script/53824928#53824928
* https://github.com/tanaikech/RunAll
* https://gist.github.com/sdesalas/2972f8647897d5481fd8e01f03122805

**Reply by johtso** [reply]
Oh, what a stupid question, sorry, don't work much with Google apps script, forgot fetch was synchronous!

**Reply by ahochsteger** [reply]
No problem, it's still a valid question I asked myself lately in another context. That's where I came across the links I posted above, although I never had a use case that would satisfy the added complexity.

**Comment by ahochsteger** [comment]
Since the original question has been answered, I'll close this discussion.
